arch: riscv: Add function to clear all unlocked PMP entries

fsammoura1980 opened the pull request:
Introduce the new function `z_riscv_pmp_clear_all()` to securely reset the Physical Memory Protection (PMP) configuration.

This function iterates through all configured PMP slots. For each entry, it checks if the Lock (L) bit is set. If the entry is not locked, it clears the Address Matching Mode (A) bits, effectively setting the region type to OFF (Null Region), disabling the entry.

The function ensures it operates in Machine mode with MSTATUS.MPRV = 0 before modifying any PMP Control and Status Registers (CSRs).

This provides a mechanism to clear all non-locked PMP regions, returning them to a default disabled state. The function is exposed in the pmp.h header file.

fsammoura1980 committed:
arch: riscv: Add z_riscv_pmp_clear_all() to reset PMP entries

Introduce the new function `riscv_pmp_clear_all()` to reset the Physical
Memory Protection (PMP) configuration.

This function iterates through all configured PMP slots. For each entry,
it writes 0x0 to the entry's 8-bit configuration register. This action
attempts to clear all fields, including the Address Matching Mode (A) bits
(setting the region type to OFF), the permission bits (R, W, X), and
the Lock (L) bit.

According to the RISC-V specification, any writes to the configuration
or address registers of a locked PMP entry are ignored. Thus, locked
entries will remain unchanged, while all non-locked entries will be
effectively disabled and their permissions cleared.

The function ensures it operates in Machine mode with MSTATUS.MPRV = 0
and MSTATUS.MPP = M-mode before modifying any PMP Control and Status
Registers (CSRs).

This provides a mechanism to clear all non-locked PMP regions,
returning them to a default disabled state. The function declaration is
exposed in the `include/zephyr/arch/riscv/pmp.h` header file, making it
available for inclusion and use by external modules.

It is recommended for firmware to call this function before transitioning
from a Read-Only (RO) stage to a Read-Write (RW) stage. This ensures
that any PMP settings established during the RO phase, which might no
longer be appropriate, are cleared, providing a clean and secure base
PMP configuration for the RW firmware.

Signed-off-by: Firas Sammoura <fsammoura@google.com>

fkokosinski reviewed (changes requested):
This would need automated testing in https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/arch/riscv/pmp

fsammoura1980 committed:
tests: riscv: Add unit tests for clearing unlocked PMP entries

Adds a new test suite to verify the behavior of `riscv_pmp_clear_all()`.
These tests ensure that the function correctly clears all unlocked PMP
entries while preserving any entries that are locked.

Signed-off-by: Firas Sammoura <fsammoura@google.com>

npitre reviewed (changes requested):
I don't like the trend I'm seeing here.

OK for having `z_riscv_pmp_read_config()` factored out. But those reading
functions are meant for debugging only. We don't want people to start
relying on those during normal operations. Then `z_riscv_pmp_write_config()`
that I disagree with, and commit 9482f8df0248 from PR #97489 should be
reverted.

The top comment in pmp.c clearly expose the strategy to follow in order for
this code to remain efficient with the lowest possible overhead during
thread context switching. The presence of `z_riscv_pmp_write_config()`
doesn't follow that strategy.

Furthermore this is pure code duplication. The same functionality already
exists. If you do want a full PMP reset function you should do this instead:

```
diff --git a/arch/riscv/core/pmp.S b/arch/riscv/core/pmp.S
index f475ce00248..eeb2cfc1706 100644
--- a/arch/riscv/core/pmp.S
+++ b/arch/riscv/core/pmp.S
@@ -93,6 +93,19 @@ pmpcfg_done:
 	add t0, t0, a0
 	jr t0
 
+/*
+ * Prototype:
+ *
+ * void z_riscv_clear_all_pmp_entries(void)
+ *
+ * Called from pmp.c to clear all PMP entries.
+ *
+ * Note: this is shared with tail clearing code above, hence the same section.
+ */
+
+GTEXT(z_riscv_clear_all_pmp_entries)
+z_riscv_clear_all_pmp_entries:
+
 pmpcfg_zerotail:
 
 	.option push
diff --git a/arch/riscv/core/pmp.c b/arch/riscv/core/pmp.c
index a3a18bd8892..e4af365e0b5 100644
--- a/arch/riscv/core/pmp.c
+++ b/arch/riscv/core/pmp.c
@@ -408,18 +408,8 @@ void riscv_pmp_clear_all(void)
 	csr_clear(mstatus, MSTATUS_MPRV);
 	csr_set(mstatus, MSTATUS_MPP);
 
-	const size_t num_pmpcfg_reg = CONFIG_PMP_SLOTS / PMPCFG_STRIDE;
-	unsigned long pmp_cfg[num_pmpcfg_reg];
-
-	z_riscv_pmp_read_config(pmp_cfg, num_pmpcfg_reg);
-
-	uint8_t *pmp_n_cfg = (uint8_t *)pmp_cfg;
-
-	for (int index = 0; index < CONFIG_PMP_SLOTS; ++index) {
-		pmp_n_cfg[index] = 0x0;
-	}
-
-	z_riscv_pmp_write_config(pmp_cfg, num_pmpcfg_reg);
+	extern void z_riscv_clear_all_pmp_entries(void);
+	z_riscv_clear_all_pmp_entries();
 }
 
 /**
```

And the global symbol for `riscv_pmp_clear_all()` should have a `z_` prefix.

Finally, I question the need for this function in the first place. You
certainly know that all unused PMP entries are already cleared in
`z_riscv_pmp_init()` where `write_pmp_entries()` is called with
`clear_trailing_entries=true`. So I'd like to see more justification
for needing a separate clear function. The "use by external modules" is
rather worrisome. If this is for when Zephyr itself is giving up control
to another software entity (if that's the RO stage vs RW stage you are
referring to) then this should be stated more generically.

fsammoura1980 commented:
> I don't like the trend I'm seeing here.
> 
> OK for having `z_riscv_pmp_read_config()` factored out. But those reading functions are meant for debugging only. We don't want people to start relying on those during normal operations. Then `z_riscv_pmp_write_config()` that I disagree with, and commit [9482f8d](https://github.com/zephyrproject-rtos/zephyr/commit/9482f8df0248f8546aab710097e7225ceeb72765) from PR #97489 should be reverted.
> 
> The top comment in pmp.c clearly expose the strategy to follow in order for this code to remain efficient with the lowest possible overhead during thread context switching. The presence of `z_riscv_pmp_write_config()` doesn't follow that strategy.
> 
> Furthermore this is pure code duplication. The same functionality already exists. If you do want a full PMP reset function you should do this instead:
> 
> ```
> diff --git a/arch/riscv/core/pmp.S b/arch/riscv/core/pmp.S
> index f475ce00248..eeb2cfc1706 100644
> --- a/arch/riscv/core/pmp.S
> +++ b/arch/riscv/core/pmp.S
> @@ -93,6 +93,19 @@ pmpcfg_done:
>  	add t0, t0, a0
>  	jr t0
>  
> +/*
> + * Prototype:
> + *
> + * void z_riscv_clear_all_pmp_entries(void)
> + *
> + * Called from pmp.c to clear all PMP entries.
> + *
> + * Note: this is shared with tail clearing code above, hence the same section.
> + */
> +
> +GTEXT(z_riscv_clear_all_pmp_entries)
> +z_riscv_clear_all_pmp_entries:
> +
>  pmpcfg_zerotail:
>  
>  	.option push
> diff --git a/arch/riscv/core/pmp.c b/arch/riscv/core/pmp.c
> index a3a18bd8892..e4af365e0b5 100644
> --- a/arch/riscv/core/pmp.c
> +++ b/arch/riscv/core/pmp.c
> @@ -408,18 +408,8 @@ void riscv_pmp_clear_all(void)
>  	csr_clear(mstatus, MSTATUS_MPRV);
>  	csr_set(mstatus, MSTATUS_MPP);
>  
> -	const size_t num_pmpcfg_reg = CONFIG_PMP_SLOTS / PMPCFG_STRIDE;
> -	unsigned long pmp_cfg[num_pmpcfg_reg];
> -
> -	z_riscv_pmp_read_config(pmp_cfg, num_pmpcfg_reg);
> -
> -	uint8_t *pmp_n_cfg = (uint8_t *)pmp_cfg;
> -
> -	for (int index = 0; index < CONFIG_PMP_SLOTS; ++index) {
> -		pmp_n_cfg[index] = 0x0;
> -	}
> -
> -	z_riscv_pmp_write_config(pmp_cfg, num_pmpcfg_reg);
> +	extern void z_riscv_clear_all_pmp_entries(void);
> +	z_riscv_clear_all_pmp_entries();
>  }
>  
>  /**
> ```
> 
> And the global symbol for `riscv_pmp_clear_all()` should have a `z_` prefix.
> 
> Finally, I question the need for this function in the first place. You certainly know that all unused PMP entries are already cleared in `z_riscv_pmp_init()` where `write_pmp_entries()` is called with `clear_trailing_entries=true`. So I'd like to see more justification for needing a separate clear function. The "use by external modules" is rather worrisome. If this is for when Zephyr itself is giving up control to another software entity (if that's the RO stage vs RW stage you are referring to) then this should be stated more generically.

We have an application that jump from RO firmware to RW firmware and I need to reset the status before jumping. This function will be called from our application. 

All read and write function are still static inline and they are not exposed. I am just going to use them for for testing purposes, such that I do not rewrite the logic. They will be only accessible for testing purposes only.

fsammoura1980 commented:
> > I don't like the trend I'm seeing here.
> > OK for having `z_riscv_pmp_read_config()` factored out. But those reading functions are meant for debugging only. We don't want people to start relying on those during normal operations. Then `z_riscv_pmp_write_config()` that I disagree with, and commit [9482f8d](https://github.com/zephyrproject-rtos/zephyr/commit/9482f8df0248f8546aab710097e7225ceeb72765) from PR #97489 should be reverted.
> > The top comment in pmp.c clearly expose the strategy to follow in order for this code to remain efficient with the lowest possible overhead during thread context switching. The presence of `z_riscv_pmp_write_config()` doesn't follow that strategy.
> > Furthermore this is pure code duplication. The same functionality already exists. If you do want a full PMP reset function you should do this instead:
> > ```
> > diff --git a/arch/riscv/core/pmp.S b/arch/riscv/core/pmp.S
> > index f475ce00248..eeb2cfc1706 100644
> > --- a/arch/riscv/core/pmp.S
> > +++ b/arch/riscv/core/pmp.S
> > @@ -93,6 +93,19 @@ pmpcfg_done:
> >  	add t0, t0, a0
> >  	jr t0
> >  
> > +/*
> > + * Prototype:
> > + *
> > + * void z_riscv_clear_all_pmp_entries(void)
> > + *
> > + * Called from pmp.c to clear all PMP entries.
> > + *
> > + * Note: this is shared with tail clearing code above, hence the same section.
> > + */
> > +
> > +GTEXT(z_riscv_clear_all_pmp_entries)
> > +z_riscv_clear_all_pmp_entries:
> > +
> >  pmpcfg_zerotail:
> >  
> >  	.option push
> > diff --git a/arch/riscv/core/pmp.c b/arch/riscv/core/pmp.c
> > index a3a18bd8892..e4af365e0b5 100644
> > --- a/arch/riscv/core/pmp.c
> > +++ b/arch/riscv/core/pmp.c
> > @@ -408,18 +408,8 @@ void riscv_pmp_clear_all(void)
> >  	csr_clear(mstatus, MSTATUS_MPRV);
> >  	csr_set(mstatus, MSTATUS_MPP);
> >  
> > -	const size_t num_pmpcfg_reg = CONFIG_PMP_SLOTS / PMPCFG_STRIDE;
> > -	unsigned long pmp_cfg[num_pmpcfg_reg];
> > -
> > -	z_riscv_pmp_read_config(pmp_cfg, num_pmpcfg_reg);
> > -
> > -	uint8_t *pmp_n_cfg = (uint8_t *)pmp_cfg;
> > -
> > -	for (int index = 0; index < CONFIG_PMP_SLOTS; ++index) {
> > -		pmp_n_cfg[index] = 0x0;
> > -	}
> > -
> > -	z_riscv_pmp_write_config(pmp_cfg, num_pmpcfg_reg);
> > +	extern void z_riscv_clear_all_pmp_entries(void);
> > +	z_riscv_clear_all_pmp_entries();
> >  }
> >  
> >  /**
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And the global symbol for `riscv_pmp_clear_all()` should have a `z_` prefix.
> > Finally, I question the need for this function in the first place. You certainly know that all unused PMP entries are already cleared in `z_riscv_pmp_init()` where `write_pmp_entries()` is called with `clear_trailing_entries=true`. So I'd like to see more justification for needing a separate clear function. The "use by external modules" is rather worrisome. If this is for when Zephyr itself is giving up control to another software entity (if that's the RO stage vs RW stage you are referring to) then this should be stated more generically.
> 
> We have an application that jump from RO firmware to RW firmware and I need to reset the status before jumping. This function will be called from our application.
> 
> All read and write function are still static inline and they are not exposed. I am just going to use them for for testing purposes, such that I do not rewrite the logic. They will be only accessible for testing purposes only.

I am fine reverting the `z_riscv_pmp_write_config()`. I only needed it for this CL and you proposed a different approach that does not require this. I updated the PR based on your recommendations.

npitre reviewed (commented):
Some minor nits:

- Similarly to `z_riscv_write_pmp_entries()` you should consider declaring
  `z_riscv_clear_all_pmp_entries()` above `z_riscv_pmp_clear_all()` rather
  than inside it, with proper documentation.

- Then I'd suggest moving the lot above the `PMP_M_MODE()` definition
  right after `write_pmp_entries()` to form a more logical grouping.

- And please move the `z_riscv_pmp_read_config()` definition on a separate
  line from the `IF_DISABLED(CONFIG_ZTEST, (static inline))`.

fsammoura1980 commented:
@npitre I responded to your comments. I believe the CQ failure is not related to my changes. I am seeing it on the main branch as well

```
./twister -v -i -p mps2/an385 -s arch.arm.user.stack -c
```

fsammoura1980 commented:
> Some minor nits:
> 
> * Similarly to `z_riscv_write_pmp_entries()` you should consider declaring
>   `z_riscv_clear_all_pmp_entries()` above `z_riscv_pmp_clear_all()` rather
>   than inside it, with proper documentation.
> * Then I'd suggest moving the lot above the `PMP_M_MODE()` definition
>   right after `write_pmp_entries()` to form a more logical grouping.
> * And please move the `z_riscv_pmp_read_config()` definition on a separate
>   line from the `IF_DISABLED(CONFIG_ZTEST, (static inline))`.

Changes made per recommendation
